File: 0731/solution.py
from sortedcontainers import SortedDict


class MyCalendarTwo:

    def __init__(self):
        self.bookings = SortedDict()

    def book(self, start: int, end: int) -> bool:
        self.bookings[start] = self.bookings.get(start, 0) + 1
        self.bookings[end] = self.bookings.get(end, 0) - 1
        count = 0
        for val in self.bookings.values():
            count += val
            if count > 2:
                self.bookings[start] = self.bookings.get(start, 0) - 1
                self.bookings[end] = self.bookings.get(end, 0) + 1
                return False
        return True

File: 0731/solution.java
import java.util.TreeMap;

class MyCalendarTwo {
    TreeMap<Integer, Integer> bookings;

    public MyCalendarTwo() {
        bookings = new TreeMap<>();
    }

    public boolean book(int start, int end) {
        bookings.put(start, bookings.getOrDefault(start, 0) + 1);
        bookings.put(end, bookings.getOrDefault(end, 0) - 1);
        int total = 0;
        for (Integer val : bookings.values()) {
            total += val;
            if (total > 2) {
                bookings.put(start, bookings.get(start) - 1);
                bookings.put(end, bookings.get(end) + 1);
                return false;
            }
        }
        return true;
    }
}

// time for book O(n)
// space O(# of bookings)
